os/StartProcess

leongross opened the pull request:
This is a rework of https://github.com/tinygo-org/tinygo/pull/4323 (based on that branch) to move the Execve and Fork wrappers to the `os` package to circumvent the goroot overrides.

leongross committed:
add StartProcess functionality (fix)

leongross committed:
renaming

leongross committed:
os rework

aykevl commented on src/os/osexec.go:
```diff
@@ -0,0 +1,56 @@
+	"unsafe"
+)
+
+func Fork() (pid int, err error) {
```
Please unexport `Fork` and `Execve` because they aren't provided by the "big Go" os package (and are easy to misuse).

aykevl commented on src/os/osexec.go:
```diff
@@ -0,0 +1,56 @@
+//go:build linux
+// +build linux
```
You only need `//go:build linux`, the old `// +build` lines are deprecated.
Also, you'll probably want `//go:build linux && !baremetal`.

aykevl commented on src/os/exec_other.go:
```diff
@@ -0,0 +1,21 @@
+//go:build !aix && !android && !freebsd && !linux && !netbsd && !openbsd && !plan9 && !solaris
+// +build !aix,!android,!freebsd,!linux,!netbsd,!openbsd,!plan9,!solaris
+
+package os
```
I think you need to define `Interrupt` and `Kill` here to get the Windows and MacOS CI to pass.

aykevl commented on src/os/exec_linux.go:
```diff
@@ -0,0 +1,84 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build posix || aix || linux || dragonfly || freebsd || (js && wasm) || netbsd || openbsd || solaris || wasip1
+// +build posix aix linux dragonfly freebsd js,wasm netbsd openbsd solaris wasip1
```
I think you can simplify this to `//go:build linux`.
Also, what about MacOS? I don't see why it wouldn't work.

aykevl reviewed (commented):
Quick initial (incomplete) review

leongross committed:
adjust build tags, un-export fork and execve, add signals for windows and macos

Signed-off-by: leongross <leon.gross@9elements.com>

leongross commented:
> main_test.go:394: /home/runner/work/tinygo/tinygo/src/os/osexec.go:14:23: undefined: syscall.Syscall

[ci](https://github.com/tinygo-org/tinygo/actions/runs/10250765856/job/28357229355?pr=4377#step:19:92) fails with undefined `syscall.Syscall`. I thought since it is just a package the compiler would pick that up when building it. @aykevl 

EDIT: it looks like this is only happening on architectures that the `createRawSyscall` function does not target, which makes sense, namely `EmulatedCortexM3 : cortex-m-qemu` and `EmulatedRISCV : riscv-qemu`. Anyways, this should not happen.
https://github.com/tinygo-org/tinygo/blob/dev/main_test.go#L522


Anyhow the smoke tests fail without any further information of what went wrong on linux, windows and mac. the smoke test cannot be removed, it is not clear though why it wasn't build in the first place

```
rm: cannot remove 'smoke.test': No such file or directory
```

leongross committed:
update wrapper comment

leongross commented:
Not sure if something like https://github.com/tinygo-org/tinygo/pull/4398 would fix this? @aykevl

aykevl commented:
> Not sure if something like #4398 would fix this? @aykevl

No, that's the wrong approach. If `syscall.Syscall` fails on baremetal systems, then the solution is not to implement these syscalls (because they're baremetal, so there is no OS to call into), but rather to figure out _why_ these calls are made.

aykevl commented on src/os/osexec.go:
```diff
@@ -0,0 +1,55 @@
+//go:build linux
```
```suggestion
//go:build linux && !baremetal
```

Baremetal systems act like Linux, so you need to explicitly exclude these.

leongross replied:
But then most of the `os` package should have that set, right? 
The CI fails for other missing implementations such as `os.Getpid` and others.
I am note sure though why this is happening here and has not happened before?

leongross committed:
exclude baremetal targets

leongross committed:
test: remove os.Getpid in seed generation on baremetal devices

leongross committed:
remove os/signal from baremetal

leongross committed:
adjust baremetal build tags

Signed-off-by: leongross <leon.gross@9elements.com>

leongross commented on src/os/tempfile.go:
```diff
@@ -17,7 +17,8 @@ var minrandMutex sync.Mutex
 	// Avoid getting same results on every run
 	now := time.Now()
-	seed := uint32(Getpid()) ^ uint32(now.Nanosecond()) ^ uint32(now.Unix())
+	// seed := uint32(Getpid()) ^ uint32(now.Nanosecond()) ^ uint32(now.Unix())
```
To get this PR working I needed to adjust the seeding function. @aykevl is this still fine or should we increase entropy somehow?

deadprogram replied:
@aykevl what do you think?

Also @leongross you should probably remove the commented out line.

aykevl replied:
Why didn't this work?

leongross replied:
I am not sure anymore what I thought the issue was there. Using the prior implementation seems to work now so I will revert this change.

leongross commented:
@deadprogram @dgryski @aykevl build and tests finally succeed, would appreciate another review.

leongross commented:
@aykevl @deadprogram @dgryski would someone mind reviewing?

deadprogram commented on src/os/osexec.go:
```diff
@@ -0,0 +1,55 @@
+	}
+	env1[len(envv)] = nil
+
+	// fail := libc_execve(&argv0[0], &argv1[0], &env1[0])
```
Any reason these lines are commented here? :smile_cat:

leongross replied:
No, this surely can be removed.

leongross committed:
remove old comments

aykevl commented on src/os/exec.go:
```diff
@@ -57,8 +61,9 @@ type Process struct {
 	Pid int
 }
 
+// Wrapper for the forkExec function, which is a wrapper around the fork and execve system calls.
```
That's not a super descriptive comment (users don't need to know that an internal function is called `forkExec`).
Maybe [`os.StartProcess`](https://pkg.go.dev/os#StartProcess) can give an idea what to put here?

aykevl commented on src/syscall/syscall_unix.go:
```diff
@@ -1,3 +1,6 @@
+//go:build linux || unix
+// +build linux unix
```
```suggestion
//go:build linux || unix
```

You can remove the old `// +build` lines.

aykevl commented on src/syscall/syscall_unix.go:
```diff
@@ -16,3 +19,8 @@ type SockaddrInet6 struct {
 	raw    RawSockaddrInet6
 }
+
+const (
+	SYS_FORK   = 57
+	SYS_EXECVE = 59
+)
```
These can be removed, because every actual operating system we have (Linux, MacOS, Windows) doesn't use the TinyGo syscall package.
(If removing these leads to compilation errors, you're probably using them from something that's not an operating system and probably need to adjust build tags at the call site).

aykevl commented on src/os/osexec.go:
```diff
@@ -0,0 +1,55 @@
+)
+
+func fork() (pid int, err error) {
+	// ret := libc_fork()
+	// ret, _, _ := syscall.Syscall(syscall.SYS_FORK, 0, 0, 0)
+	ret, _, _ := syscall.Syscall(57, 0, 0, 0)
```
Don't use magic numbers. Use `syscall.SYS_FORK` instead, for example.

leongross replied:
I removed the explicit syscall number and found out, that in the CI the definition for `SYS_FORK` cannot be found. Grepping through the source tree I found out, that there is no definition for this, only the local go installation provides this, e.g. via `src/syscall/zsysnum_linux_amd64.go`. What confused me is that the CI should have picked this up but did not. On my local machine, this works perfectly fine.

aykevl commented on src/os/exec_other.go:
```diff
@@ -0,0 +1,28 @@
+//go:build (!aix && !android && !freebsd && !linux && !netbsd && !openbsd && !plan9 && !solaris) || baremetal || tinygo.wasm
+// +build !aix,!android,!freebsd,!linux,!netbsd,!openbsd,!plan9,!solaris baremetal tinygo.wasm
```
```suggestion
```

`// +build` lines can be removed.

aykevl commented on src/os/osexec.go:
```diff
@@ -0,0 +1,55 @@
+
+	// fail := libc_execve(&argv0[0], &argv1[0], &env1[0])
+	// fail, _, _ := syscall.Syscall(syscall.SYS_EXECVE, uintptr(unsafe.Pointer(&argv0[0])), uintptr(unsafe.Pointer(&argv1[0])), uintptr(unsafe.Pointer(&env1[0])))
+	fail, _, _ := syscall.Syscall(59, uintptr(unsafe.Pointer(&argv0[0])), uintptr(unsafe.Pointer(&argv1[0])), uintptr(unsafe.Pointer(&env1[0])))
```
Please don't use magic numbers. It looks like you can use `syscall.Exec` instead.

aykevl commented on src/os/osexec.go:
```diff
@@ -0,0 +1,55 @@
+	// fail, _, _ := syscall.Syscall(syscall.SYS_EXECVE, uintptr(unsafe.Pointer(&argv0[0])), uintptr(unsafe.Pointer(&argv1[0])), uintptr(unsafe.Pointer(&env1[0])))
+	fail, _, _ := syscall.Syscall(59, uintptr(unsafe.Pointer(&argv0[0])), uintptr(unsafe.Pointer(&argv1[0])), uintptr(unsafe.Pointer(&env1[0])))
+	if fail < 0 {
+		// TODO: parse the syscall return codes
+		return errors.New("execve failed")
```
Can you implement this TODO?
(If you use `syscall.Exec` you already have an error value to return!)

aykevl commented on src/os/exec_linux.go:
```diff
@@ -0,0 +1,83 @@
+// The StartProcess function creates a new process by forking the current process and then calling execve to replace the current process with the new process.
+// It thereby replaces the newly created process with the specified command and arguments.
+// Differences to upstream golang implementation (https://cs.opensource.google/go/go/+/master:src/syscall/exec_unix.go;l=143):
+// * No setting of Process Attributes
+// * Ignoring Ctty
+// * No ForkLocking (might be introduced by #4273)
+// * No parent-child communication via pipes (TODO)
+// * No waiting for crashes child processes to prohibit zombie process accumulation / Wait status checking (TODO)
```
Is it possible to check for these cases, and if users use any of these to return an error instead of continuing and not doing what the user expected?
For example, if `Stdin` is set, it should either implement that feature or return an error.

leongross replied:
That sounds like a reasonable thing to do. I will evaluate this and see how/if this is possible.

aykevl commented on src/os/exec.go:
```diff
@@ -47,6 +47,10 @@ func (p *ProcessState) Sys() interface{} {
 	return nil // TODO
 }
 
+func (p *ProcessState) Exited() bool {
+	return false // TODO
+}
```
That's... a little ugly.
Is there any way we can track whether the process is still running?

leongross replied:
For now, I stubbed it out the same way all the other methods over `ProcessState` (which is an empty struct) are implemented. Making this functional would mean a quite large rework of the tinygo process representation, which atm is just a pid. I think it is a good idea to rework this, but [not as part of this PR](https://github.com/tinygo-org/tinygo/issues/4471).

aykevl replied:
Oh I see, `Wait` doesn't return anything really.

aykevl reviewed (commented):
* There is some commented out code, that looks like testing code. IMHO it should either be used, or removed (it doesn't really add any value in documentation).
* You make a few changes to the syscall package, but why? All actual operating systems (Linux, macOS, Windows) use the default syscall package. If you need any changes to the syscall package, that probably indicates that it is called from somewhere (for example, baremetal) and you need to adjust build tags.

leongross committed:
os/StartProcess: cleanup

remove old comments, replace explicit syscall numbers with named
variants, undo tempfile seed generation, remove old // +build tags,
improve StartProcess documentation, remove custom syscall number
definitions.

Signed-off-by: leongross <leon.gross@9elements.com>

deadprogram commented:
@leongross still are some CI failures from this PR is would appear.

leongross commented:
I figured out the error here: there is no fork syscall on arm64 [1]. The original golang implementation abstracts this better and does not do raw syscalls as my implementation does. So for now, arm64 has to be excluded as a possible target architecture for this.

[1]: https://chromium.googlesource.com/chromiumos/docs/+/master/constants/syscalls.md#cross_arch-numbers

leongross committed:
os/exec: exclude aarch64

aarch64 has no support for the fork ssycall. for now exclude it.

Signed-off-by: leongross <leon.gross@9elements.com>

leongross committed:
add fallback stub for aarch64

leongross committed:
add syscall error handling

Signed-off-by: leongross <leon.gross@9elements.com>

leongross committed:
change build tags aarch64 -> arm64

Signed-off-by: leongross <leon.gross@9elements.com>

leongross committed:
add signal implementations for linux arm64

leongross committed:
wip

leongross commented:
Windows CI fails on, any will investigate.
```
--- FAIL: TestBuild/Host/go1.21.go (35.61s)
```

aykevl commented:
I think you need to drop the first commit, now that you've rebased on the dev branch?

The Windows test failures are probably just because the Windows CI tends to be flaky (sometimes, tests don't produce any output and we don't know why) so you can ignore these for now.

leongross commented:
Thanks for the feedback @aykevl. It looks like now just the said windows CI is failing.

leongross commented:
@aykevl any idea how we can fix the flaky windows CI?

aykevl commented on src/syscall/syscall_unix.go:
```diff
@@ -1,3 +1,5 @@
+//go:build linux || unix
```
```suggestion
//go:build unix
```

aykevl commented on src/os/osexec.go:
```diff
@@ -0,0 +1,56 @@
+//go:build linux && !baremetal && !darwin && !tinygo.wasm && !arm64
```
```suggestion
//go:build linux && !baremetal && !tinygo.wasm && !arm64
```

It can't be both linux and darwin.

aykevl commented on src/os/osexec.go:
```diff
@@ -0,0 +1,56 @@
+	"unsafe"
+)
+
+func fork() (pid int, err error) {
+	ret, _, err := syscall.Syscall(syscall.SYS_FORK, 0, 0, 0)
+	if int(ret) != 0 {
+		errno := err.(syscall.Errno)
+		if int(errno) != 0 {
+			return -1, errno
+		}
+	}
+	return int(ret), nil
+}
```
Can you please either implement `fork()` for arm64, or just use the libc function?

For example (this code is entirely untested but something like that should work):

```suggestion
func fork() (pid int32, err error) {
	pid = libc_fork()
	if pid != 0 {
		err = syscall.Errno(*libc_errno())
	}
	return
}

//export fork
func libc_fork() int32

// Internal musl function to get the C errno pointer.
//export __errno_location
func libc_errno() *int32
```

leongross replied:
Thanks for the input, this seems to be a better idea. I had a look at the execve implementation, but I think there it does not make sense to call into musl since it is the exact same thing we do here:

```c
int execve(const char *path, char *const argv[], char *const envp[])
{
	return syscall(SYS_execve, path, argv, envp);
}

```

aykevl commented on src/os/exec_linux_test.go:
```diff
@@ -0,0 +1,32 @@
+		return
+	}
+
+	proc, err := StartProcess("/bin/echo", []string{"hello", "world"}, &ProcAttr{})
```
Can you also add a negative test, to make sure that running `/bin/non-existing-program` for example returns the expected error message?

Can you also add some tests that verify that unsupported features like `ProcAttr.Dir`, `ProcAttr.Sys`, etc will return an error as they should? (These checks probably also need to be implemented). If a program sets `Dir` for example but then the program executes in another directory without error, that's going to be weird and difficult to debug. I'd rather not have users deal with such bugs.

aykevl commented on src/syscall/exec_unix.go:
```diff
@@ -0,0 +1,7 @@
+
+package syscall
+
+// Implemented in runtime package.
+func runtime_BeforeExec()
+func runtime_AfterExec()
```
Looks like these are never called? Why did you declare them?

leongross replied:
These were leftover stubs back when https://github.com/tinygo-org/tinygo/pull/4426 was not merged yet.

aykevl commented on src/os/osexec.go:
```diff
@@ -0,0 +1,56 @@
+
+	ret, _, err := syscall.Syscall(syscall.SYS_EXECVE, uintptr(unsafe.Pointer(&argv0[0])), uintptr(unsafe.Pointer(&argv1[0])), uintptr(unsafe.Pointer(&env1[0])))
+	if int(ret) != 0 {
+		// errno := err.(syscall.Errno)
+		// return errno
```
Please clean up these kinds of leftover comments.

aykevl commented on src/os/signal/signal.go:
Please remove this file.

aykevl reviewed (commented):
Reviewed.

A number of comments I made are basically TODO items left in the code that really should have been implemented in some way, or just stuff that isn't needed and should have been removed from this PR. When submitting a PR, can you take a look at it to make sure you implement/cleanup/remove as much of this as possible? That makes reviewing a lot faster.

As for the Windows CI, that's a long standing issue and unrelated to this PR. I don't know how to fix it (I'm not running Windows myself and it's difficult and time consuming to fix issues like those).

leongross committed:
update TestForkExec with invalid testcase

leongross committed:
remove comments, remove unused/obsolete files

Signed-off-by: leongross <leon.gross@9elements.com>

leongross committed:
replace raw syscall implementation with musl wrapper

Signed-off-by: leongross <leon.gross@9elements.com>

leongross committed:
remove arm64 handling

Signed-off-by: leongross <leon.gross@9elements.com>

leongross committed:
update musl builder to include arch specific process files

Signed-off-by: leongross <leon.gross@9elements.com>

leongross commented:
The CI seems to have problems finding musl files that are there although. Locally running these commands does not produce errors. @aykevl do you have any idea what is going on there? Only the smoke tests fail, the unit tests seem to work out.


https://github.com/tinygo-org/tinygo/actions/runs/11590162170/job/32267411673?pr=4377#step:10:11
https://github.com/tinygo-org/tinygo/actions/runs/11590162170/job/32267411673?pr=4377#step:9:700

leongross committed:
add musl sources to release files

Signed-off-by: leongross <leon.gross@9elements.com>

leongross commented:
@aykevl

b0ch3nski commented on src/os/exec_linux.go:
```diff
@@ -0,0 +1,103 @@
+
+	if err != nil {
+		return 0, err
+	} else {
```
nitpick: redundant `else`

leongross committed:
remove redundant else

aykevl commented on src/os/exec_linux.go:
```diff
@@ -0,0 +1,103 @@
+		return 0, errors.New("exec: no argv")
+	}
+
+	if len(argv) == 0 {
```
Nitpick: with this check, the previous check is unnecessary. `len(argv)` of a nil slice is always 0.

b0ch3nski replied:
I think this was left behind and got merged :slightly_smiling_face:

aykevl replied:
Yeah, it wasn't a blocker for me. It's not a bug, just code that could have been cleaner.

aykevl commented on src/os/os.test:
This file needs to be removed.

aykevl commented on src/os/osexec_aarch64.go:
```diff
@@ -0,0 +1,18 @@
+//go:build linux && !baremetal && !darwin && !tinygo.wasm && aarch64
```
`aarch64` is not a valid Go build tag.

I think this file can be removed entirely now, since these things are implemented in osexec.go?

aykevl reviewed (commented):
Just a few small things, otherwise this PR seems ready for merging.

aykevl commented on src/os/osexec_aarch64.go:
```diff
@@ -0,0 +1,18 @@
+//go:build linux && !baremetal && !darwin && !tinygo.wasm && aarch64
+
+// arm64 does not have a fork syscall, so ignore it for now
+// TODO: add support for arm64 with clone or use musl implementation
```
This comment needs to be updated or removed since it is outdated.

aykevl commented on src/os/osexec.go:
```diff
@@ -0,0 +1,61 @@
+//go:build linux && !baremetal && !tinygo.wasm
+
+// arm64 does not have a fork syscall, so ignore it for now
+// TODO: add support for arm64 with clone or use musl implementation
```
This comment needs to be updated or removed since it is outdated.

aykevl commented:
Thank you! I have merged the PR.

leongross commented:
Thanks for merging and the thorough review @aykevl!
